lib/report_generator.rb

```diff
@@ -72,5 +59,22 @@ def build_report_pms(query, table = "", usernames = [])
       end
       pms
     end
+
+    private
+
+    def filter_recipients_by_query_access(recipients, query)
+      recipients.reduce([]) do |names, recipient|
+        if (group = Group.find_by(name: recipient)) &&
+             (
+               group.id == Group::AUTO_GROUPS[:admins] ||
+                 query.query_groups.exists?(group_id: group.id)
+             )
+          next names.concat group.users.pluck(:username)
+        elsif (user = User.find_by(username: recipient))
+          next names << recipient if Guardian.new(user).user_can_access_query?(query)
+        end
+        next names
```

Comment:
```suggestion
      recipients.reduce([]) do |names, recipient|
        if (group = Group.find_by(name: recipient)) &&
             (
               group.id == Group::AUTO_GROUPS[:admins] ||
                 query.query_groups.exists?(group_id: group.id)
             )
          names.concat group.users.pluck(:username)
        elsif (user = User.find_by(username: recipient))
          names << recipient if Guardian.new(user).user_can_access_query?(query)
        else
          names
        end
```

Reply:
Your line 74 will result in a `names` becoming `nil` if the condition does not satisfy.

Reply:
Haha saw the edit. I will leave it as is.

Reply:
Yea the key with `reduce` is that you have to return the state at the end of every iteration. I think that is what was causing the problem in the first place.